// eslint-plugin-antfu@0.19.4/dist/index.mjs
import { ESLintUtils } from '@typescript-eslint/utils';

const createEslintRule = ESLintUtils.RuleCreator((ruleName) => ruleName);

const RULE_NAME$1 = "if-newline";
const ifNewline = createEslintRule({
  name: RULE_NAME$1,
  meta: {
    type: "problem",
    docs: {
      description: "Newline after if",
      recommended: "error"
    },
    fixable: "code",
    schema: [],
    messages: {
      missingIfNewline: "Expect newline after if"
    }
  },
  defaultOptions: [],
  create: (context) => {
    return {
      IfStatement(node) {
        if (!node.consequent)
          return;
        if (node.consequent.type === "BlockStatement")
          return;
        if (node.test.loc.end.line === node.consequent.loc.start.line) {
          context.report({
            node,
            loc: {
              start: node.test.loc.end,
              end: node.consequent.loc.start
            },
            messageId: "missingIfNewline",
            fix(fixer) {
              return fixer.replaceTextRange([node.consequent.range[0], node.consequent.range[0]], "\n");
            }
          });
        }
      }
    };
  }
});

const RULE_NAME = "no-leading-newline";
const noLeadingNewline = createEslintRule({
  name: RULE_NAME,
  meta: {
    type: "problem",
    docs: {
      description: "Do not allow leading newline",
      recommended: "error"
    },
    fixable: "whitespace",
    schema: [],
    messages: {
      noLeadingNewline: "No leading newline allowed"
    }
  },
  defaultOptions: [],
  create: (context) => {
    return {
      Program(node) {
        const code = context.getSourceCode();
        const match = code.text.match(/^[\s]+/)?.[0] || "";
        if (match.includes("\n")) {
          const line = match.split("\n");
          context.report({
            node,
            loc: {
              start: { line: 0, column: 0 },
              end: { line: line.length - 1, column: line[line.length - 1].length }
            },
            messageId: "noLeadingNewline",
            fix(fixer) {
              return fixer.replaceTextRange([0, match.length], "");
            }
          });
        }
      }
    };
  }
});

const index = {
  rules: {
    "no-leading-newline": noLeadingNewline,
    "if-newline": ifNewline
  }
};

export { index as default };
